worlds/marioland2/__init__.py

```diff
@@ -0,0 +1,576 @@
+
+    web = MarioLand2WebWorld()
+
+    item_name_groups = {
+        "Level Progression": {item_name for item_name in items if item_name.endswith(("Progression", "Secret",
+                                                                                      "Secret 1", "Secret 2"))
+                              and "Auto Scroll" not in item_name},
+        "Bells": {item_name for item_name in items if "Bell" in item_name},
+        "Golden Coins": {"Mario Coin", "Macro Coin", "Space Coin", "Tree Coin", "Turtle Coin", "Pumpkin Coin"},
+        "Coins": {"1 Coin", *{f"{i} Coins" for i in range(2, 169)}},
+        "Powerups": {"Mushroom", "Fire Flower", "Carrot"},
+        "Difficulties": {"Easy Mode", "Normal Mode"},
+        "Auto Scroll Traps": {item_name for item_name in items
+                              if "Auto Scroll" in item_name and "Cancel" not in item_name},
+        "Cancel Auto Scrolls": {item_name for item_name in items if "Cancel Auto Scroll" in item_name},
+    }
+
+    location_name_groups = {
+        "Bosses": {
+            "Tree Zone 5 - Boss", "Space Zone 2 - Boss", "Macro Zone 4 - Boss",
+            "Pumpkin Zone 4 - Boss", "Mario Zone 4 - Boss", "Turtle Zone 3 - Boss"
+                   },
+        "Normal Exits": {location for location in locations if locations[location]["type"] == "level"},
+        "Secret Exits": {location for location in locations if locations[location]["type"] == "secret"},
+        "Bells": {location for location in locations if locations[location]["type"] == "bell"},
+        "Coins": {location for location in location_name_to_id if "Coin" in location}
+    }
+
+    options_dataclass = SML2Options
+    options: SML2Options
+
+    generate_output = generate_output
+
+    def __init__(self, world, player: int):
+        super().__init__(world, player)
+        self.auto_scroll_levels = []
+        self.num_coin_locations = []
+        self.max_coin_locations = {}
+        self.sprite_data = {}
+        self.coin_fragments_required = 0
+
+    def generate_early(self):
+        self.sprite_data = deepcopy(level_sprites)
+        if self.options.randomize_enemies:
+            randomize_enemies(self.sprite_data, self.random)
+        if self.options.randomize_platforms:
+            randomize_platforms(self.sprite_data, self.random)
+
+        if self.options.marios_castle_midway_bell:
+            self.sprite_data["Mario's Castle"][35]["sprite"] = "Midway Bell"
+
+        if self.options.auto_scroll_chances == -1:
+            self.auto_scroll_levels = [int(i in [19, 25, 30]) for i in range(32)]
+        else:
+            self.auto_scroll_levels = [int(self.random.randint(1, 100) < self.options.auto_scroll_chances)
+                                       for _ in range(32)]
+
+        self.auto_scroll_levels[level_name_to_id["Mario's Castle"]] = 0
+        unbeatable_scroll_levels = ["Tree Zone 3", "Macro Zone 2", "Space Zone 1", "Turtle Zone 2", "Pumpkin Zone 2"]
```

Comment:
You missed this one
```suggestion
        unbeatable_scroll_levels = ["Tree Zone 3", "Macro Zone 2", "Space Zone 1", "Turtle Zone 2", "Pumpkin Zone 2", "Space Zone Secret Course"]
```

Reply:
I have beaten this level with auto scroll active several times. I don't understand. It's a very straightforward level.
![image](https://github.com/user-attachments/assets/6b765523-7864-46aa-b32f-0fb34afb90d1)

Reply:
Then the logic for it is wrong, since it is causing errors in generations